src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
+The Authorization, Crypto-Key , and TTL headers take a little explanation. Let's 
+start with the simplest.
+
+## HTTP headers
 
-_endpoint_—Contains two parts: the URL of the messaging service you're using
-followed by unique identifier for the user. This is called a subscription ID or
-a registration ID. This tells your server how to identify you the messaging
-server.
+### TTL
 
-_keys_—Encryption keys used for encrypting data passed to the service worker
-messages.
+It may be some time before a message server can deliver a message sent by your 
+app server. Message services aren't obligated to hold a message forever. 
+Frankly, to be timely, an app server should never send a message that can live 
+forever. That's why it's required to include a header called TTL, literally 
+'time to live'.
 
-## Encrypting a message {#encrypting}
+The TTL header is a a value in seconds given as a suggestion to the message 
+server for how long the message should be retained. If it chooses, the message 
+server may retain the message for less time than is given in the TTL header. If 
+it does, it must return that time in a TTL header in a response to the message 
+request. 
 
-You must encrypt all messages before sending them. Encryption is enough of a
-specialty, even within software development, that we don't recommend writing
-your own encryption system. Fortunately, there are [a range of push
-libraries](https://github.com/web-push-libs) including our own [Push Encryption
-Library](https://github.com/GoogleChrome/web-push-encryption).
+You can almost think of this interaction as a conversation. 
 
-## Two ways to send a push message {#ways-to-send}
+App server: "Here's a message can you deliver it within 43200 seconds?"
 
-When implementing server code there are two methods of sending messages to
-consider.
+Message server: "I really can't, but I can do it for you in 21600."
+
+If the TTL has a value of 0, the message server should deliver it immediately or 
+as soon as soon as the user agent is available to receive it.
+
+### Crypto-Key header
+
+To validate a message sent by your app server, the message server needs the 
+public key. You send the public key in the Crypto-Key header. It has several 
+parts.  
+  
+<pre>dh=<i>applicationServerKey</i>;p256ecdsa=<i>publicKey</i></pre>  
+  
+For example:  
+  
+<pre>dh=BGEw2wsHgLwzerjvnMTkbKrFRxdmwJ5S\_k7zi7A1coR\_sVjHmGrlvzYpAT1n4NPbioFlQkIrT  
+NL8EH4V3ZZ4vJE,p256ecdsa=BDd3\_hVL9fZi9Ybo2UUzA284WG5FZR30\_95YeZJsiApwXKpNcF1rRPF3foIiBHXRdJI2Qhumhf6\_LFTeZaN</pre>
+  
+The first part (`dh=applicationServerKey`) is the public key, the one we created 
+under Requesting permission and subscribing users. The second part 
+(`p256ecdsa=publicKey`) is the public key, but base64 url encoded. Notice the 
+comma separating the two parts of the crypto-key.  
+  
+**Note:** A bug in Chrome 52 requires that a semicolon be send to the message 
```

Comment:
"A bug in Chrome 52 means that you need to seperate the dh=<application server key> and p256ecdsa=<public key> sections by a semicolon rather than a comma.